test/org/apache/catalina/core/TestApplicationHttpRequest.java

```diff
@@ -354,4 +364,11 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp)
             }
         }
     }
+
+    private static boolean objectsEqual(String object1, String object2) {
```

Comment:
`java.util.Objects.equals(Object a, Object b)` ?

Reply:
That is problematic if this will be backported to Tomcat 7.

Reply:
The code here fails if only the first parameter is `null`

Reply:
I agree that `getParameter` should remain: a parameter present with no value should return `""` and not `null`.

Reply:
Indeed. I did not use `Objects.equals` because of backward compatibility.
I will solve that possible NPE.